satellite/gracefulexit/endpoint.go

```diff
@@ -509,71 +408,55 @@ func (endpoint *Endpoint) processIncomplete(ctx context.Context, stream processS
-			nodePiece = piece
-		}
-		excludedNodeIDs[i] = piece.NodeId
-	}
-
-	if nodePiece == nil {
-		endpoint.log.Debug("piece no longer held by node", zap.Stringer("Node ID", nodeID), zap.ByteString("path", incomplete.Path), zap.Int32("piece num", incomplete.PieceNum))
 
-		err = endpoint.db.DeleteTransferQueueItem(ctx, nodeID, incomplete.Path, incomplete.PieceNum)
+	// validate pointer state
+	nodePiece, err := validatePointer(ctx, pointer, incomplete)
+	if err != nil {
+		endpoint.log.Debug("piece no longer held by node", zap.Stringer("node ID", nodeID), zap.ByteString("path", incomplete.Path), zap.Int32("piece num", incomplete.PieceNum))
+		err := endpoint.db.DeleteTransferQueueItem(ctx, nodeID, incomplete.Path, incomplete.PieceNum)
 		if err != nil {
 			return Error.Wrap(err)
 		}
+		return Error.Wrap(err)
+	}
 
+	pieceSize, err := endpoint.calculatePieceSize(ctx, pointer, incomplete, nodePiece)
+	if ErrAboveOptimalThreshold.Has(err) {
 		return nil
 	}
-
-	redundancy, err := eestream.NewRedundancyStrategyFromProto(pointer.GetRemote().GetRedundancy())
 	if err != nil {
 		return Error.Wrap(err)
 	}
 
-	if len(remote.GetRemotePieces()) > redundancy.OptimalThreshold() {
-		endpoint.log.Debug("pointer has more pieces than required. removing node from pointer.", zap.Stringer("Node ID", nodeID), zap.ByteString("path", incomplete.Path), zap.Int32("piece num", incomplete.PieceNum))
-
-		_, err = endpoint.metainfo.UpdatePieces(ctx, string(incomplete.Path), pointer, nil, []*pb.RemotePiece{nodePiece})
-		if err != nil {
-			return Error.Wrap(err)
-		}
-
-		err = endpoint.db.DeleteTransferQueueItem(ctx, nodeID, incomplete.Path, incomplete.PieceNum)
-		if err != nil {
-			return Error.Wrap(err)
-		}
-
-		return nil
+	// populate excluded node IDs
+	pieces := pointer.GetRemote().RemotePieces
+	excludedNodeIDs := make([]storj.NodeID, len(pieces))
+	for i, piece := range pieces {
+		excludedNodeIDs[i] = piece.NodeId
 	}
 
-	pieceSize := eestream.CalcPieceSize(pointer.GetSegmentSize(), redundancy)
-
-	request := overlay.FindStorageNodesRequest{
+	// get replacement node
+	request := &overlay.FindStorageNodesRequest{
 		RequestedCount: 1,
 		FreeBandwidth:  pieceSize,
 		FreeDisk:       pieceSize,
 		ExcludedNodes:  excludedNodeIDs,
 	}
 
-	newNodes, err := endpoint.overlay.FindStorageNodes(ctx, request)
+	newNodes, err := endpoint.overlay.FindStorageNodes(ctx, *request)
 	if err != nil {
 		return Error.Wrap(err)
 	}
 
 	if len(newNodes) == 0 {
 		return Error.New("could not find a node to receive piece transfer: node ID %v, path %v, piece num %v", nodeID, incomplete.Path, incomplete.PieceNum)
 	}
+
 	newNode := newNodes[0]
 	endpoint.log.Debug("found new node for piece transfer", zap.Stringer("original node ID", nodeID), zap.Stringer("replacement node ID", newNode.Id),
 		zap.ByteString("path", incomplete.Path), zap.Int32("piece num", incomplete.PieceNum))
 
+	remote := pointer.GetRemote()
```

Comment:
maybe you can do this up before you call `pointer.GetRemote().RemotePieces` above. That way you can have
```
remote := pointer.GetRemote()
pieces := remote.RemotePieces
```